File: 3965-earliest-finish-time-for-land-and-water-rides-i/earliest-finish-time-for-land-and-water-rides-i.py
class Solution:
    def earliestFinishTime(self, landStartTime: List[int], landDuration: List[int], waterStartTime: List[int], waterDuration: List[int]) -> int:
        ans = float('inf')

        for i in range(len(landStartTime)):
            for j in range(len(waterStartTime)):

                # Land -> Water
                land_finish = landStartTime[i] + landDuration[i]
                water_start = max(
                    land_finish,
                    waterStartTime[j]
                )

                ans = min(
                    ans,
                    water_start + waterDuration[j]
                )

                # Water -> Land
                water_finish = (
                    waterStartTime[j]
                    + waterDuration[j]
                )

                land_start = max(
                    water_finish,
                    landStartTime[i]
                )

                ans = min(
                    ans,
                    land_start + landDuration[i]
                )

        return ans

File: 3965-earliest-finish-time-for-land-and-water-rides-i/earliest-finish-time-for-land-and-water-rides-i.cpp
class Solution {
public:
    int earliestFinishTime(vector<int>& landStartTime, vector<int>& landDuration, vector<int>& waterStartTime, vector<int>& waterDuration) {
        int ans = INT_MAX;

        for (int i = 0; i < landStartTime.size(); i++) {
            for (int j = 0; j < waterStartTime.size(); j++) {

                // Land -> Water
                int landFinish =
                    landStartTime[i] + landDuration[i];

                int waterStart =
                    max(landFinish,
                        waterStartTime[j]);

                ans = min(
                    ans,
                    waterStart + waterDuration[j]
                );

                // Water -> Land
                int waterFinish =
                    waterStartTime[j] +
                    waterDuration[j];

                int landStart =
                    max(waterFinish,
                        landStartTime[i]);

                ans = min(
                    ans,
                    landStart + landDuration[i]
                );
            }
        }

        return ans;
        
    }
};